### DISCUSSION TITLE
Redirect URI set correctly but getting "Incorrect redirect_uri" error from token endpoint

### DISCUSSION THREAD

**Discussion by dheeraj-326** [keycloak/keycloak#13661] [Q&A] [answered]
Redirect URIs are set correctly
![image](https://user-images.githubusercontent.com/24998581/183584578-4fe69a00-62d0-455b-8a21-ecc0cfe1c0eb.png)
But I am getting this error while trying to get the access_token
![image](https://user-images.githubusercontent.com/24998581/183584670-96446fbb-d092-460c-9206-9fe7b47b8b33.png)

Please HELP! :(

**Comment by mhajas** [comment]
@dheeraj-326 Where and when did you obtain the `code` parameter? 

It seems you are using [Authorization code flow](https://www.keycloak.org/docs/latest/server_admin/index.html#_oidc-auth-flows-authorization). The request you added shows step 6. Is it possible that step 2 was done with a different redirect_uri than step 6?

**Reply by dheeraj-326** [reply]
Thank you for your response. No, I have used the same redirect_uri on the previous step as well.

**Reply by mhajas** [reply]
The error is thrown here: https://github.com/keycloak/keycloak/blob/e44cea587fe97d73b83de2ea8e943cf079ce95ee/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L370-L373

You can try to put a breakpoint there and see what is the difference between `redirectUriParam` (the one from your request) and `redirectUri` (the one stored in step 2).

**Comment by dheeraj-326** [comment]
Thanks for that. Locally debugging Keycloak might indeed be my only option
here.

**Comment by dheeraj-326** [comment, marked as answer]
I found the solution. The configuration I have set in Keycloak configuration is "http://locahost:8000/*". But my "redirect_uri" was "http://locahost:8000". The issue was resolved when I added "http://localhost:8000" to "Valid redirect URIs" in client config.

So it seems that adding an asterisk at the end is equivalent to a '+' in regex where it expects one or more items.

**Reply by dheeraj-326** [reply]
@mhajas
